esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5892,260 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+    if (!df)
+        throw MakeStringException(ECLWATCH_FILE_NOT_EXIST,"Cannot find file %s.", req.getName());
+
+    CFileAccessRole role = req.getAccessRole();
+
+    switch (role)
+    {
+        case CFileAccessRole_Token:
+        {
+            break;
+        }
+        case CFileAccessRole_Engine:
+        {
+            /* JCSMORE - for now do nothing
+             * Ideally, would get the file tree here and add it to 'metaInfo' tree, i.e. outside of uncrypted secureInfo blob
+             * Then client could construct a IDistributeFile from it etc.
+             * However, the way the engines and IDistributedFile work at the moment, means that using this info
+             * at the client side would require a significant amount of refactoring of the IDistributedFile implementation.
+             * Not least because IDF allows updates via IPT -> Dali.
+             *
+             * So for now don't send anything, and rely on engine fetching the legacy way, i.e. direct from Dali, via lazy fetching etc.
+             */
+            break;
+        }
+        case CFileAccessRole_External:
+        {
+            IArrayOf<IEspDFUPartLocations> dfuPartLocations;
+            IArrayOf<IEspDFUPartCopies> dfuPartCopies;
+            resp.setNumParts(getFilePartsInfo(context, df, req.getCluster(), dfuPartLocations, dfuPartCopies));
+            resp.setFilePartLocations(dfuPartLocations);
+            resp.setFileParts(dfuPartCopies);
+            if (req.getReturnJsonTypeInfo() || req.getReturnJsonTypeInfo())
+            {
+                MemoryBuffer binLayout;
+                StringBuffer jsonLayout;
+                if (!getRecordFormatFromRtlType(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo()))
+                    getRecordFormatFromECL(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo());
+                if (req.getReturnJsonTypeInfo() && jsonLayout.length())
+                    resp.setRecordTypeInfoJson(jsonLayout.str());
+                if (req.getReturnBinTypeInfo() && binLayout.length())
+                    resp.setRecordTypeInfoBin(binLayout);
+            }
+            break;
+        }
+        default:
+            throwUnexpected();
+    }
+
+    // setup "expiryTime"
+    unsigned expirySecs = req.getExpirySeconds();
+    if (expirySecs > fileAccessExpirySeconds)
+        expirySecs = fileAccessExpirySeconds;
+    time_t now;
+    time(&now);
+    CDateTime expiryDt;
+    expiryDt.set(now + expirySecs);
+    StringBuffer expiryTimeStr;
+    expiryDt.getString(expiryTimeStr);
+
+    StringBuffer keyPairName;
+    getFileDafilesrvKeyName(keyPairName, *df);
+
+    Owned<IPropertyTree> metaInfo = createPTree();
+    getFileMeta(*metaInfo, *df, udesc, role, expiryTimeStr, keyPairName, req);
+    StringBuffer metaInfoJsonBlob;
+    toJSON(metaInfo, metaInfoJsonBlob);
```

Comment:
Do we always return the MetaInfoBlob in json format? ESP response can be in JSON (for url http://ip:port/WsDfu/DFUFileAccess.json) or in XML (http://ip:port/WsDfu/DFUFileAccess).

Reply:
yes, the format is not for esp clients, but for dafilesrv.
Esp clients should treat it as an opaque blob.